tools/upload_firmware.py

```diff
@@ -39,10 +39,17 @@
                 with open(args.bin_file, 'rb') as bin_file:
                     firmware_image = bin_file.read()
 
-                success = client.writeFlash([board_id], args.offset, firmware_image, sector_map=flash_sector_maps, print_progress=True)
+                success = False
+                while not success:
+                    try:
+                        success = client.writeFlash([board_id], args.offset, firmware_image, sector_map=flash_sector_maps, print_progress=True)
+                    except (MalformedPacketError, ProtocolError) as e:
+                        print(f'Upload to board {board_id} failed with error:')
+                        print(e)
+                        print('Retrying...')
+                        continue
 
                 if success:
-                    # client.leaveBootloader(board_id)
                     print("Upload to Board", board_id, "Succeeded")
                 else:
                     print("Upload to Board", board_id, "Failed")
```

Comment:
it seems like this will never reach the else statement anymore?

Reply:
Yup. I'm debating whether to add a retry limit though so idk.